wrapper/src/main/java/software/amazon/jdbc/MonitorPluginService.java

```diff
@@ -0,0 +1,795 @@
+    this.props = props;
+    this.originalUrl = originalUrl;
+    this.driverProtocol = targetDriverProtocol;
+    this.targetDriverDialect = targetDriverDialect;
+    this.dbDialect = dbDialect;
+    this.configurationProfile = configurationProfile;
+    this.exceptionManager = exceptionManager;
+    this.connectionProviderManager = new ConnectionProviderManager(
+        this.pluginManager.getDefaultConnProvider(),
+        this.pluginManager.getEffectiveConnProvider());
+
+    this.sessionStateService = sessionStateService != null
+        ? sessionStateService
+        : new SessionStateServiceImpl(this, this.props);
+
+    this.exceptionHandler = this.configurationProfile != null && this.configurationProfile.getExceptionHandler() != null
+        ? this.configurationProfile.getExceptionHandler()
+        : null;
+  }
+
+  @Override
+  public Connection getCurrentConnection() {
+    return this.currentConnection;
+  }
+
+  @Override
+  public HostSpec getCurrentHostSpec() {
+    if (this.currentHostSpec == null) {
+      this.currentHostSpec = this.initialConnectionHostSpec;
+
+      if (this.currentHostSpec == null) {
+        if (this.getAllHosts().isEmpty()) {
+          throw new RuntimeException(Messages.get("PluginServiceImpl.hostListEmpty"));
+        }
+
+        this.currentHostSpec = this.getWriter(this.getAllHosts());
+        final List<HostSpec> allowedHosts = this.getHosts();
+        if (!Utils.containsUrl(allowedHosts, this.currentHostSpec.getUrl())) {
+          throw new RuntimeException(
+              Messages.get("PluginServiceImpl.currentHostNotAllowed",
+                  new Object[] {
+                      currentHostSpec == null ? "<null>" : currentHostSpec.getUrl(),
+                      Utils.logTopology(allowedHosts, "")})
+          );
+        }
+
+        if (this.currentHostSpec == null) {
+          this.currentHostSpec = this.getHosts().get(0);
+        }
+      }
+      if (this.currentHostSpec == null) {
+        throw new RuntimeException("Current host is undefined.");
+      }
+      LOGGER.finest(() -> "Set current host to " + this.currentHostSpec);
+    }
+    return this.currentHostSpec;
+  }
+
+  public void setInitialConnectionHostSpec(final @NonNull HostSpec initialConnectionHostSpec) {
+    this.initialConnectionHostSpec = initialConnectionHostSpec;
+  }
+
+  @Override
+  public HostSpec getInitialConnectionHostSpec() {
+    return this.initialConnectionHostSpec;
+  }
+
+  @Override
+  public String getOriginalUrl() {
+    return this.originalUrl;
+  }
+
+  @Override
+  public ServiceContainer getServiceContainer() {
+    return this.serviceContainer;
+  }
+
+  @Override
+  @Deprecated
+  public void setAllowedAndBlockedHosts(AllowedAndBlockedHosts allowedAndBlockedHosts) {
+    this.storageService.set(this.initialConnectionHostSpec.getHost(), allowedAndBlockedHosts);
+  }
+
+  @Override
+  public boolean acceptsStrategy(HostRole role, String strategy) throws SQLException {
```

Comment:
Supporting strategies (acceptStrategy and getHostSpecByStrategy) is interesting. It's not too hard to support them but we kinda not expecting forceConnect pipeline to use them. Can we just temporary let them raise an exception?